File: src/main/scan.js
'use strict';
/** 扫描文件夹，把 RAW+JPG 同名文件合并成一组 */
const fsp = require('fs').promises;
const path = require('path');
const { isSupported, isRaw } = require('./preview');

const SKIP_DIRS = new Set(['$RECYCLE.BIN', 'System Volume Information', 'node_modules', '.git', 'Lightroom', '.thumbnails']);

const collator = new Intl.Collator('zh-Hans-CN', { numeric: true, sensitivity: 'base' });

async function walk(dir, out, opts, depth) {
  let entries;
  try {
    entries = await fsp.readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const e of entries) {
    const full = path.join(dir, e.name);
    if (e.isDirectory()) {
      if (!opts.recursive) continue;
      if (depth >= 8) continue;
      if (e.name.startsWith('.') || SKIP_DIRS.has(e.name)) continue;
      if (opts.excludeDirs.some((d) => path.resolve(d).toLowerCase() === path.resolve(full).toLowerCase())) continue;
      await walk(full, out, opts, depth + 1);
    } else if (e.isFile()) {
      if (e.name.startsWith('.')) continue;
      if (!isSupported(full)) continue;
      out.push(full);
    }
  }
}

/**
 * @param {string} root
 * @param {{recursive:boolean, group:boolean, excludeDirs:string[]}} opts
 */
async function scanFolder(root, opts) {
  const files = [];
  await walk(root, files, opts, 0);

  const map = new Map();
  for (const f of files) {
    const dir = path.dirname(f);
    const base = path.basename(f, path.extname(f));
    const key = opts.group ? `${dir.toLowerCase()}|${base.toLowerCase()}` : f.toLowerCase();
    let g = map.get(key);
    if (!g) {
      g = { dir, base, files: [] };
      map.set(key, g);
    }
    g.files.push(f);
  }

  const groups = [];
  for (const g of map.values()) {
    g.files.sort(); // 先定序，下面的 sizes 要和 files 一一对应
    // 主文件：优先 RAW（内嵌预览质量最高且带完整 EXIF）
    const raws = g.files.filter(isRaw);
    const primary = raws.length ? raws[0] : g.files[0];
    let stat;
    try {
      stat = await fsp.stat(primary);
    } catch {
      continue;
    }
    let total = 0;
    const sizes = [];
    for (const f of g.files) {
      let s = 0;
      try {
        s = (await fsp.stat(f)).size;
      } catch {}
      sizes.push(s);
      total += s;
    }
    groups.push({
      id: path.relative(root, primary) || path.basename(primary),
      dir: g.dir,
      base: g.base,
      name: path.basename(primary),
      files: g.files,
      sizes,
      primary,
      ext: path.extname(primary).slice(1).toUpperCase(),
      isRaw: raws.length > 0,
      hasPair: g.files.length > 1,
      size: total,
      mtime: stat.mtimeMs,
    });
  }

  groups.sort((a, b) => collator.compare(a.id, b.id));
  return groups;
}

module.exports = { scanFolder };

File: src-tauri/src/scan.rs
use crate::fav::read_favorites;
use crate::types::{PhotoGroup, ScanResult};
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};
use std::time::UNIX_EPOCH;
use walkdir::WalkDir;

const RAW_EXTS: &[&str] = &[
    "nef", "nrw", "cr2", "cr3", "crw", "arw", "srf", "sr2", "dng", "raf", "orf", "rw2", "pef",
    "ptx", "srw", "raw", "rwl", "iiq", "3fr", "fff", "erf", "mrw", "x3f",
];

const PLAIN_EXTS: &[&str] = &["jpg", "jpeg", "jpe", "png", "webp", "gif", "bmp", "avif"];

const SKIP_DIRS: &[&str] = &["$RECYCLE.BIN", "System Volume Information", "node_modules", ".git"];

pub fn is_raw_ext(ext: &str) -> bool {
    let lower = ext.to_lowercase();
    RAW_EXTS.contains(&lower.as_str())
}

pub fn is_supported_ext(ext: &str) -> bool {
    let lower = ext.to_lowercase();
    RAW_EXTS.contains(&lower.as_str()) || PLAIN_EXTS.contains(&lower.as_str())
}

/// 遍历时顺手记下来的文件信息，省掉后面每个文件再 stat 一次
struct FileInfo {
    path: PathBuf,
    size: u64,
    mtime: f64,
}

pub fn scan_folder(
    root: &str,
    recursive: bool,
    group_raw_jpeg: bool,
    fav_file_name: &str,
) -> ScanResult {
    let root_path = Path::new(root);
    if !root_path.is_dir() {
        return ScanResult {
            ok: false,
            root: root.to_string(),
            groups: Vec::new(),
            favorites_file_name: fav_file_name.to_string(),
        };
    }

    let max_depth = if recursive { 8 } else { 1 };
    let mut files: Vec<FileInfo> = Vec::new();

    for entry in WalkDir::new(root_path)
        .max_depth(max_depth)
        .into_iter()
        // filter_entry 也会作用于根节点：根目录自己叫 ".photos" 之类时不能把整棵树否掉
        .filter_entry(|e| {
            if e.depth() == 0 {
                return true;
            }
            let name = e.file_name().to_string_lossy();
            !(name.starts_with('.') || SKIP_DIRS.iter().any(|&d| name.eq_ignore_ascii_case(d)))
        })
        .filter_map(|e| e.ok())
    {
        if !entry.file_type().is_file() {
            continue;
        }
        let p = entry.path();
        let supported = p
            .extension()
            .and_then(|x| x.to_str())
            .map(is_supported_ext)
            .unwrap_or(false);
        if !supported {
            continue;
        }
        // Windows 上 walkdir 直接复用 readdir 拿到的元数据，这里不会真的再打一次盘
        let md = match entry.metadata() {
            Ok(m) => m,
            Err(_) => continue,
        };
        files.push(FileInfo {
            path: p.to_path_buf(),
            size: md.len(),
            mtime: md
                .modified()
                .ok()
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_millis() as f64)
                .unwrap_or(0.0),
        });
    }

    // 分组：同目录同主名的 RAW / JPG 归为一组
    let mut map: HashMap<String, Vec<FileInfo>> = HashMap::new();
    for f in files {
        let key = if group_raw_jpeg {
            let dir = f
                .path
                .parent()
                .unwrap_or(root_path)
                .to_string_lossy()
                .to_lowercase();
            let stem = f.path.file_stem().unwrap_or_default().to_string_lossy().to_lowercase();
            format!("{}|{}", dir, stem)
        } else {
            f.path.to_string_lossy().to_lowercase()
        };
        map.entry(key).or_default().push(f);
    }

    // 收藏列表
    let fav_items = read_favorites(root, fav_file_name);
    let fav_set: HashSet<String> = fav_items.iter().map(|x| x.trim().to_lowercase()).collect();

    let rel_of = |p: &Path| -> Option<String> {
        pathdiff::diff_paths(p, root_path).map(|r| r.to_string_lossy().replace('\\', "/"))
    };

    let mut groups = Vec::new();
    for mut file_list in map.into_values() {
        file_list.sort_by(|a, b| a.path.cmp(&b.path)); // 确定顺序
        if file_list.is_empty() {
            continue;
        }

        // 主文件优先选 RAW
        let primary_idx = file_list
            .iter()
            .position(|f| {
                f.path
                    .extension()
                    .and_then(|x| x.to_str())
                    .map(is_raw_ext)
                    .unwrap_or(false)
            })
            .unwrap_or(0);
        let has_raw = file_list.iter().any(|f| {
            f.path
                .extension()
                .and_then(|x| x.to_str())
                .map(is_raw_ext)
                .unwrap_or(false)
        });
        let primary = &file_list[primary_idx];

        let primary_name = primary.path.file_name().unwrap_or_default().to_string_lossy().to_string();
        let rel_id = rel_of(&primary.path).unwrap_or_else(|| primary_name.clone());

        // 收藏可以按相对路径 / 文件名 / 主名任一种写法命中
        let id_lower = rel_id.to_lowercase();
        let name_lower = primary_name.to_lowercase();
        let base_lower = primary
            .path
            .file_stem()
            .unwrap_or_default()
            .to_string_lossy()
            .to_lowercase();
        let is_fav = fav_set.contains(&id_lower)
            || fav_set.contains(&name_lower)
            || fav_set.contains(&base_lower)
            || file_list.iter().any(|f| {
                let n = f.path.file_name().unwrap_or_default().to_string_lossy().to_lowercase();
                let r = rel_of(&f.path).map(|x| x.to_lowercase()).unwrap_or_default();
                fav_set.contains(&n) || (!r.is_empty() && fav_set.contains(&r))
            });

        groups.push(PhotoGroup {
            id: rel_id,
            dir: primary.path.parent().unwrap_or(root_path).to_string_lossy().to_string(),
            base: primary.path.file_stem().unwrap_or_default().to_string_lossy().to_string(),
            name: primary_name,
            files: file_list.iter().map(|f| f.path.to_string_lossy().to_string()).collect(),
            sizes: file_list.iter().map(|f| f.size).collect(),
            primary: primary.path.to_string_lossy().to_string(),
            ext: primary
                .path
                .extension()
                .unwrap_or_default()
                .to_string_lossy()
                .to_uppercase(),
            is_raw: has_raw,
            has_pair: file_list.len() > 1,
            size: file_list.iter().map(|f| f.size).sum(),
            mtime: primary.mtime,
            favored: is_fav,
        });
    }

    // 自然排序（按 id）
    groups.sort_by(|a, b| alphanumeric_sort::compare_str(&a.id, &b.id));

    ScanResult {
        ok: true,
        root: root.to_string(),
        groups,
        favorites_file_name: fav_file_name.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn raw_and_plain_extensions() {
        assert!(is_raw_ext("nef"));
        assert!(is_raw_ext("NEF"));
        assert!(is_raw_ext("CR3"));
        assert!(!is_raw_ext("jpg"));
        assert!(is_supported_ext("jpg"));
        assert!(is_supported_ext("AVIF"));
        assert!(!is_supported_ext("txt"));
        assert!(!is_supported_ext(""));
    }

    #[test]
    fn missing_root_is_not_ok() {
        let r = scan_folder("Z:/nope/nope", true, true, "favorites.txt");
        assert!(!r.ok);
        assert!(r.groups.is_empty());
    }
}
